### DISCUSSION TITLE
Strangely high volume_other_latency

### DISCUSSION THREAD

**Discussion by rodenj1** [NetApp/harvest#2335] [Q&A]
I am trying to reconcile some odd differences between Active IQ performance Stats and Harvest collected stats.  I am comparing volume_read_latency, volume_write_latency, and volume_other_latency to the "Volume / Performance - Latency" in Active IQ.  I know this will not be perfect match, but I would like to assume that it would be in the ballpark.  And maybe I need to be using a different set of harvest stats to compare against.

Here is what I am seeing...

"Other" ops stats seem off by quite a bit
volume_other_latency = **163ms**
Active IQ Volume Other Latency = **1ms**
volume_other_ops = 5IOPS
 Active IQ Volume Other Ops = 1IOPS

"Write" ops stats seem close
volume_write_latency = 9.2ms
Active IQ Volume Write Latency = 9.2ms
volume_write_ops = 3791IOPS
 Active IQ Volume Write Ops = 2088

"Reads" ops stats seem close
volume_read_latency = 0.6ms
Active IQ Volume Read Latency = 0.4ms
volume_read_ops = 1122IOPS
 Active IQ Volume Read Ops = 1743IOPS

Here is what I am running...
Harvest: 2023.08.0
Using: zapi
Cluster Ontap: 9.11.1P8

I am thinking that volume_*_latency and volume_*_ops includes other operations that Active IQ is not.  If so, is there a better volume level metric I could be using that would match closer to what Active IQ is showing?

**Comment by rahulguptajss** [comment]
@rodenj1 We have validated Harvest volume performance metrics against ONTAP. Could you try steps mentioned [here](https://github.com/NetApp/harvest/wiki/Volume-latency-for-flexgroup) and let us know if you see differences.

**Comment by rodenj1** [comment]
Thank you for providing the doc.  I have verified that the general volume_avg_latency matches to the cli.  It is the volume_other_latency that is way off.  And there does not seem to be a CLI command that will show that too me that I can compare against.  I have pulled the raw api data and followed the formula in the document of `avg_latency = delta(avg latency raw)/ delta(iops raw)` and came up with very similar results to what volume_other_latency shows.  So it does not seem to be a Harvest math issue, it feels more like the cluster is not calculating correctly.

Here is volume_other_latency
![volume_other_latency](https://github.com/NetApp/harvest/assets/8315258/7badbab9-5451-4a91-ad82-74ba24cbfe94)

Here is me replicating it with the raw data from the API using this formula `other latency = delta(other latency raw)/ delta(other iops raw)`.  Note, I did this with "delta" and "rate" and it was very similar
![volume_other_latency_raw](https://github.com/NetApp/harvest/assets/8315258/efaf80e6-999b-4537-af19-3b150ef34fe8)

Here is what Active IQ shows for Volume Other Latency.  The thing to note here is volume other latency is below 1ms for the 60 min period.
![oum_other_latency](https://github.com/NetApp/harvest/assets/8315258/a16f1081-f668-45c2-8e80-76a2159a9b45)

At this point I think I have verified that Harvest is correctly displaying what the cluster it telling it.  It seems that the cluster might not displaying the metric correct.  Would you agree with that?  Can you think of any other way of verifying this stat?

**Comment by rahulguptajss** [comment]
@rodenj1 Based on the screenshots you've provided, here are the observations.

Harvest provides volume_other_latency, which is measured in microseconds, as detailed in the Harvest [documentation](https://netapp.github.io/harvest/23.08/ontap-metrics/#volume_other_latency). From the screenshot you provided, it appears that the 'volume_other_latency' reported by Harvest is approximately 300 microseconds, which is equivalent to 0.3 milliseconds.

On the other hand, the AIQ, as per your screenshot, reports a latency of 0.6 milliseconds. Although there is a slight difference between these two values, they are relatively close. This discrepancy could be attributed to factors such as variations in the collection schedule, timing, and other parameters.

Could you confirm if that understanding is correct or highlight any unclear discrepancies.

**Reply by rodenj1** [reply]
Hi Rahul, all of the latency_other screen shots all include a "/1000" which is already converting from microseconds to millisecond.  So it is reporting 300 milliseconds.

**Reply by rahulguptajss** [reply]
Right. I suggest reviewing the [KB article](https://kb.netapp.com/onprem/ontap/Performance/Why_does_the_statistics_volume_show_command_show_the_total_ops_which_donot_match_the_read%2C_write%2C_and_other_ops) if it helps. If the issue appears to originate from ONTAP, I recommend reaching out to their support team for further assistance.

**Comment by rodenj1** [comment]
So, I know this is an old question, but I still have no answer for this.  I have tried to look at this in many different ways.  The KB article was not really helpful.  At this point because of all the anomalies the team no longer has any faith in the numbers coming from Harvest.  Which is sad because it is a great tool.  But I get it.  They get alerts from Harvest of high latency, they see HUGE spikes in Harvest data, but when they look at System Manager and AIQUM, these spikes don't exist and worse is the numbers just don't seem match up.  I have tried averaging out the spike, changing the sample rate and looked at different latency stats.  The issue is they just don't line up and not even close.  Here are more examples of volume latency numbers.

Harvest volume_avg_latency stat
![volume_avg_latency_raw](https://github.com/NetApp/harvest/assets/8315258/9d4d0d11-9b54-4b9d-92a8-c49c76d7aff4)

Here is the same stat with 30 min sample to match what I think I am getting from System Manager and Active IQ Unified Manager
![volume_avg_latency_30min_sampl](https://github.com/NetApp/harvest/assets/8315258/217c4085-0c8d-4c7e-ab35-26798e24c850)
Here is what system manager says
![volume_avg_latency_system_manager](https://github.com/NetApp/harvest/assets/8315258/601b4679-d15f-4377-9632-fe260ad1fae5)
Here is what AIQUM says
![volume_avg_latency_aiqum](https://github.com/NetApp/harvest/assets/8315258/f9f9294f-2833-4f26-83f6-8735d148e9fe)

Here is another volume
volume_avg_latency stat
![volume_avg_latency_raw_1](https://github.com/NetApp/harvest/assets/8315258/ef6331e5-cb56-44c8-b91f-d3830f09b9a9)
30 min sample
![volume_avg_latency_30min_sampl_1](https://github.com/NetApp/harvest/assets/8315258/3610eb72-7592-44ad-8992-72b3be35bf82)
System Manager
![volume_avg_latency_system_manager_1](https://github.com/NetApp/harvest/assets/8315258/cfa0dca7-4f4d-42f0-bc57-68a57ae2bb13)
AIQUM
![volume_avg_latency_aiqum_1](https://github.com/NetApp/harvest/assets/8315258/25fe62f9-77db-4287-b6dd-af43ff446256)


I have many more examples of this as well, things with longer spikes as well if that helps.

**Comment by rahulguptajss** [comment]
Thanks @rodenj1 for the detailed data. Since our discussion last time, a bug has been discovered in Harvest that is the same as the one you are reporting. It was fixed via #2544 and the fix is currently available in the [nightly](https://github.com/NetApp/harvest/releases/tag/nightly) builds. There is another ongoing issue in this area, #2605, which should be resolved soon.

**Reply by rodenj1** [reply]
Sounds good.  I have upgrade all my collectors to the nightly release.  I will let you know what I see.

**Reply by rahulguptajss** [reply]
@rodenj1 The fix for issue #2605 has been included in the nightly build (24.02.01-nightly). Please update to this version for all fixes related to latency spikes.

**Reply by rodenj1** [reply]
Sounds good.  I updated the fleet of collectors this morning.  Things were looking pretty good as of the last patch, so I expect more of the same.

**Reply by cgrinds** [reply]
Thanks for letting us know @rodenj1 sounds promising. Please let us know if you see anymore anomalies.

**Reply by rahulguptajss** [reply]
@rodenj1 Could you please update if the latency data looks alright now with the fix?

**Reply by rodenj1** [reply]
I have been watching for a few days now and all the anomalies have subsided.  It looks like the the fixes has resolved the issues I was seeing.  Thank you!

**Reply by cgrinds** [reply]
Awesome! Thanks for raising this issue @rodenj1 and letting us know it is resolved

**Reply by rahulguptajss** [reply]
@rodenj1 Stable release with these fixes is now available. https://github.com/NetApp/harvest/releases/tag/v24.02.0
